### ISSUE TITLE
Support snak and reference hashes

### ISSUE THREAD

**Issue by wetneb** [Wikidata/Wikidata-Toolkit#319]
The data model should be able to store reference and snak hashes, when they are provided.
This is necessary if we want to implement granular editing actions which need to identify parts of statements.

Hashes should not be required to construct the relevant objects as they are omitted in some JSON payloads (such as the dumps).

**Comment by Tanish-Gupta** [comment]
Hi, 
Is this issue still open for contribution? I am looking to contribute to it. 
From my understanding, we pass references to makeStatement() in Datamodel. I looked at the JSON for a statement, and there's one field called mainsnak and another called snak (and hash as well) within references. Can you confirm if we are looking to store the parameter - List<Reference> references passed to the makeStatement()?

If yes, do we intend to make store these as a data member of Datamodel class itself?

Thank you

**Comment by Tanish-Gupta** [comment]
> I do not understand this question, sorry. I cannot parse it.
There is a parameter List<reference> passed to makeStatement() method in Datamodel class, so my question was that do we need to store it? But, your next statement made it clear. Thank you

So, I have come up with the following solution - 
We can have a string variable "hash" and include it in the json only when it is not Null, in both ReferenceImpl and SnakImpl classes 

![image](https://user-images.githubusercontent.com/23396490/83158529-832dd380-a122-11ea-932b-29d0f1b59f7d.png)

Then. create a deserialize them 

![image](https://user-images.githubusercontent.com/23396490/83158827-e9b2f180-a122-11ea-821d-a10da1413be7.png)

![image](https://user-images.githubusercontent.com/23396490/83159043-2f6fba00-a123-11ea-95f1-a50ba02d595d.png)

This way the hashes are stored only when they are provided.

Can you please confirm if this fulfils the requirement ?

Thank you

**Comment by wetneb** [comment]
Yes this looks great to me! By the way, the best way to get feedback on your code changes is to submit them as a pull request (rather than including them as screenshots). Pull requests do not have to be final, you can always make changes to the pull request after review, or even start a new one from scratch if the direction was inappropriate.
